In [2]:
import numpy as np
import scipy.linalg as sla

#Questão 1

In [6]:
def elimin_gaussiana(a, b, imp=False, norma='inf'):
    # matriz estendida Me
    b = np.mat(b).T
    Me = np.concatenate((a, b), axis=1)
    if imp: print('Matriz estendida: \n', Me, '\n\n')
    if imp: print('=' * 80)
    for j in range(Me.shape[-1] - 2):  # coluna
        if imp: print('Operação na coluna', j)
        if imp: print('=' * 80, '\n\n')
        for i in range(j + 1, Me.shape[0]):  # linha
            Me[i] = Me[i] - (Me[i, j] / Me[j, j]) * Me[j]  # operações com cada linha
            if imp: print('Operação na linha', i, ':', Me[i], '\n')
            if imp: print('Matriz escalonando: \n', Me, '\n')
        if imp: print('=' * 80)
    if imp: print('\nMatriz escalonada: \n', Me, '\n')
    Matriz_A = np.delete(Me, -1, axis=1)  # matriz de coeficientes escalonada
    Vetor_b = Me[:, -1]  # matriz de termos independentes escalonada
    Vetor_x = sla.solve_triangular(Matriz_A, Vetor_b)  # solução do sistema escalonado
    if imp: print('=' * 80, '\n\nSolução do sistema: \n', Vetor_x)
    # verificação do erro do sistema
    if norma is 'inf': norma = np.inf
    if norma is '-inf': norma = -np.inf
    Erro = sla.norm(b - a.dot(Vetor_x), ord=norma)
    return (Matriz_A, Vetor_b, Vetor_x, Erro)


In [4]:
a = np.array([[2,3,-1],[4,4,-3],[2,-3,1]])
b = np.array([5,3,-1])
M = elimin_gaussiana(a,b)
M

(matrix([[ 2,  3, -1],
         [ 0, -2, -1],
         [ 0,  0,  5]]), matrix([[ 5],
         [-7],
         [15]]), array([[1.],
        [2.],
        [3.]]), 0.0)

In [10]:
a = np.array([[1, 1, 1], [4, 4, 2], [2, 1, -1]]).astype(float)
b = np.array([1, 2, 0]).astype(float)

#Tive que adicionar o tipo pra minha função funcionar



In [7]:
elimin_gaussiana(a, b)

/home/marlon/venv/lib/python3.7/site-packages/ipykernel_launcher.py:11: RuntimeWarning: divide by zero encountered in long_scalars
  # This is added back by InteractiveShellApp.init_path()
/home/marlon/venv/lib/python3.7/site-packages/ipykernel_launcher.py:11: RuntimeWarning: invalid value encountered in multiply
  # This is added back by InteractiveShellApp.init_path()


LinAlgError: singular matrix: resolution failed at diagonal 1

O erro ocorre pois depois da primeira subtração, a segunda linha passa a ter os valores [0, 0, -2, -2], assim, no momento em que houver a subtração da segunda terceira linha, teremos zero como pivô.

In [7]:
def elim_gaussiana(A, b, pivotamento=None, imp=False, norma='inf'):
    b = np.mat(b).T
    me = np.array(np.concatenate((A, b), axis=1))
    n = len(me)

    if imp:
        print('Matriz estendida:\n{}\n'.format(me))
    if imp:
        print('=' * 80)

    # Eliminação
    for k in range(n - 1):
        if pivotamento:
            # Pivô
            maxindex = abs(me[k:, k]).argmax() + k
            if me[maxindex, k] == 0:
                raise ValueError("Matriz é singular.")
            # Trocas
            if maxindex != k:
                me[[k, maxindex]] = me[[maxindex, k]]

                if imp:
                    print("Troca entre as linhas {} e {}".format(k, maxindex))
                if imp:
                    print('Matriz escalonando:\n{}\n'.format(me))
        else:
            if me[k, k] == 0:
                raise ValueError("Pivô é zero. Tente com a variável \'pivoteamento\' como True.")

        for linha in range(k + 1, n):
            mult = me[linha, k] / me[k, k]
            me[linha, k:] = me[linha, k:] - mult * me[k, k:]
            if imp:
                print('Operação na linha {}: {}'.format(linha, me[linha]))
            if imp:
                print('Matriz escalonando: \n {} \n'.format(me))

    # Retrosubstituição

    if imp:
        print('Matriz escalonada:\n{}\n'.format(me))

    matriz_a = me[:, :-1]  # matriz de coeficientes escalonada
    vetor_b = me[:, -1]  # matriz de termos independentes escalonada
    vetor_x = sla.solve_triangular(matriz_a, vetor_b)  # solução do sistema escalonado

    if imp:
        print('=' * 80, '\n\nSolução do sistema: \n', vetor_x)

    # Verificação do erro do sistema
    if norma is 'inf':
        norma = np.inf
    if norma is '-inf':
        norma = -np.inf
    erro = sla.norm(b - A.dot(vetor_x), ord=norma)

    return matriz_a, vetor_b, vetor_x, erro


In [11]:
elim_gaussiana(a, b, pivotamento=True)


(array([[ 4. ,  4. ,  2. ],
        [ 0. , -1. , -2. ],
        [ 0. ,  0. ,  0.5]]),
 array([ 2. , -1. ,  0.5]),
 array([ 1., -1.,  1.]),
 3.0)

#Questão 2

In [18]:
x = np.array([1., 1., 1.])
x1 = np.array([4.5, 0, 1.01])

erro_a = x - x1
erro_r = erro_a/x1
print("Erro absoluto: {}\nErro releativo: {}".format(erro_a, erro_r))

Erro absoluto: [-3.5   1.   -0.01]
Erro releativo: [-0.77777778         inf -0.00990099]


#Questão 3#

In [12]:
a = np.array([[1., 4, 52], [27, 110, -3], [22, 2, 14]])
b = np.array([57., 134, 38])

In [13]:
elim_gaussiana(a, b, pivotamento=True)


(array([[ 27.        , 110.        ,  -3.        ],
        [  0.        , -87.62962963,  16.44444444],
        [  0.        ,   0.        ,  52.09721048]]),
 array([134.        , -71.18518519,  52.09721048]),
 array([1., 1., 1.]),
 173.0)

Como podemos ver, o resultado com pivoteamento parcial é o correto, ou seja, não houve erro.

#Questão 4#
Já implementada na questão 1

#Questão 6#
Os resultados são diferentes pois o método lu retorna as matrizes L e U separadas, permitindo outras aplicações desse resultado sem a necessidade de manipulações nas matrizes.

Já o método lu_factor() retorna ambas as matrizes L e U unidas em uma só. Ou seja, no formato ideal para ser usado como entrada para a função lu_solve.

#Questão 7#

###a)
<br>
####Sala 1:
<br>Wfumantes + Qa.ca - Qa.c1 + E13(c3-c1) = 0
<br>1000 + 400 - 200c1 + 25c3 - 25c1 = 0
<br>1400 - 225c1 + 25c3 = 0
<br>225c1 - 25c3 = 1400
<br>
####Sala 2:
<br>Qb.cb+(Qa/2).c4 - Qc.c2 + E24(c4-c2) = 0
<br>50*2 +100.c4 - 150.c2 + 25.c4 - 25c2 = 0
<br>100 +100.c4 - 150.c2 + 25.c4 - 25c2 = 0
<br>175c2 - 125c4 = 100
<br>
####Sala 3:
<br>Wgrelha + Qa.c1 + E13(c1 - c3) + E34(c4-c3) - Qa.c3 = 0
<br>2000 + 200.c1 + 25c1 - 25c3 + 50.c4 - 50.c3 - 200.c3 = 0
<br>225.c1 - 275.c3 + 50.c4 = -2000
<br>
####Sala 4:
<br>Qa.c3 + E43(c3-c4)+E42(c2-c4)-Qd.c4-Qd.c4 = 0
<br>200.c3 + 50.c3 - 50.c4 + 25.c2 - 25.c4 - 200.c4 = 0
<br>25.c2 + 250.c3 - 275.c4 = 0


Transformando os dados acima em matriz:

In [57]:
mat_a = np.array([[225,0,-25,0], [0, 175, 0, -125], [225, 0, -275, 50], [0, 25, 250, -275]])
mat_b = np.array([1400,100, -2000, 0])
resultado = elim_gaussiana(mat_a, mat_b, pivotamento=True)[2]
resultado

array([ 8.09983897, 12.35196687, 16.89855072, 16.49275362])

###b)

Para saber a quantidade de cada fator na área das crianças, devemos ter a matriz b contendo apenas os valores correspondentes ao peso desse fator. <br>Ex: para os fumantes, temos:<br>b = [1000,0,0,0]



In [27]:
mat_b = np.array([1000,0,0,0])
res_i = elim_gaussiana(mat_a, mat_b, pivotamento=True)[2]
res_i


array([4.99624262, 3.45065562, 4.96618357, 4.83091787])

In [29]:
res_i[1]/resultado[1]


0.2793608224382612

####i)<br>
Como nosso interesse é apenas na sala 2, temos que dos 12.35 mg/h de monóxido de carbono que chegam a sala 2, 3.45 são devido aos fumantes, correspondendo à cerca de 27.94% da poluição na sala das crianças.

In [31]:
mat_b = np.array([0,0,-2000,0])12
res_ii = elim_gaussiana(mat_a, mat_b, pivotamento=True)[2]
res_ii

array([1.10359635, 6.90131125, 9.93236715, 9.66183575])

In [32]:
res_ii[2]/resultado[2]

0.5877644368210407

####ii)<br>
A grelha corresponde a cerca de 58.77% da poluição da sala das crianças.

In [34]:
mat_b = np.array([400,100,0,0])
res_iii = elim_gaussiana(mat_a, mat_b, pivotamento=True)[2]
res_iii

array([2., 2., 2., 2.])

In [36]:
res_iii[3]/resultado[3]

0.12126537785588752

####iii)
<br>A entrada de ventilação Qa corresponde a aproximadamente 12.13% da poluição da sala das crianças


###c)<br>

Primeiro, vamos atualizar os valores da matriz:

In [43]:
mat_a = np.array([[225, 0, -25, 0], [0, 175, 0, -125], [225, 0, -275, 50], [0, 25, 250, -275]])
mat_b = np.array([2400, 100, -5000, 0])


Agora, precisamos da inversa de A:

In [46]:
inv_a = sla.inv(mat_a)

In [47]:
res_aum = np.dot(inv_a, mat_b)
res_aum

array([14.75095785, 26.13793103, 36.75862069, 35.79310345])

In [48]:
res_aum-resultado

array([ 6.65111889, 13.78596416, 19.86006997, 19.30034983])

Assim, temos que o aumento na concentração de monóximo de carbono na sala das crianças foi de ~13.78 mg/h

####d)

In [59]:
mat_a = np.array([[225, 0, -25, 0], [0, 155, 0, -105], [225, 0, -275, 50], [0, 5, 250, -255]])
mat_b = np.array([1400, 100, -2000, 0])
res_rede = elim_gaussiana(mat_a, mat_b, pivotamento=True)[2]


In [60]:
res_rede - resultado

array([ 0.00972433, -0.23785999,  0.08751893,  0.43759464])

Concluímos então que a instalação da rede, reduziria a concentração de monóxido de carbono na sala das crianças em apenas 0.2376 mg/h

###Questão 8